gemini/function-calling/parallel_function_calling.ipynb

```diff
@@ -0,0 +1,858 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "SbmM4z7FOBpM"
+   },
+   "source": [
+    "<div class=\"alert alert-block alert-warning\">\n",
+    "<b>⚠️ The kernel is going to restart. Please wait until it is finished before continuing to the next step. ⚠️</b>\n",
+    "</div>\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "dmWOrTJ3gx13"
+   },
+   "source": [
+    "### Authenticate your notebook environment (Colab only)\n",
+    "\n",
+    "If you are running this notebook on Google Colab, run the cell below to authenticate your environment."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {
+    "id": "NyKGtVQjgx13",
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "import sys\n",
+    "\n",
+    "if \"google.colab\" in sys.modules:\n",
+    "    from google.colab import auth\n",
+    "\n",
+    "    auth.authenticate_user()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "DF4l8DTdWgPY"
+   },
+   "source": [
+    "### Set Google Cloud project information and initialize Vertex AI SDK\n",
+    "\n",
+    "To get started using Vertex AI, you must have an existing Google Cloud project and [enable the Vertex AI API](https://console.cloud.google.com/flows/enableapi?apiid=aiplatform.googleapis.com).\n",
+    "\n",
+    "Learn more about [setting up a project and a development environment](https://cloud.google.com/vertex-ai/docs/start/cloud-environment)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {
+    "id": "Nqwi-5ufWp_B",
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "PROJECT_ID = \"[your-project-id]\"  # @param {type:\"string\"}\n",
+    "LOCATION = \"us-central1\"  # @param {type:\"string\"}\n",
+    "\n",
+    "import vertexai\n",
+    "\n",
+    "vertexai.init(project=PROJECT_ID, location=LOCATION)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "jXHfaVS66_01"
+   },
+   "source": [
+    "## Code Examples\n",
+    "\n",
+    "### Import libraries"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {
+    "id": "lslYAvw37JGQ",
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "import google_play_scraper\n",
+    "from IPython.display import display, Markdown\n",
+    "from vertexai.generative_models import (\n",
+    "    FunctionDeclaration,\n",
+    "    GenerativeModel,\n",
+    "    GenerationConfig,\n",
+    "    Part,\n",
+    "    Tool,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Define helper function"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Helper function to extract one or more function calls from a Gemini Function Call response\n",
+    "def extract_function_calls(response):\n",
```

Comment:
The Vertex AI SDK has been updated so that you don't have to go as nested to access function calls.

I updated the official code samples to show this:

https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/function-calling#text-samples

Reply:
Done. Thanks for letting me know about this, it is convenient! If it returned Pythonic object types, then I could remove my helper function completely, but one step at a time. 😁